File: PTJ4-Shared-Class-Port/solution/cpp/BankAccount.cpp
#include "BankAccount.h"

#include <utility>


/*****************
*   FUNCTIONS   *
*****************/

// Initialize the owner and starting balance
BankAccount::BankAccount(std::string owner, double balance)
	: owner_(std::move(owner)), balance_(balance) {}

// Add money to the stored balance
void BankAccount::deposit(double amount) {
	balance_ += amount;
}

// Withdraw money only when the account has enough balance
bool BankAccount::withdraw(double amount) {
	// Reject withdrawals that exceed the current balance
	if (amount > balance_) {
		return false;
	}

	balance_ -= amount;
	return true;
}

// Build a readable account summary
std::string BankAccount::summary() const {
	return owner_ + " has $" + std::to_string(balance_);
}

File: PTJ4-Shared-Class-Port/solution/java/BankAccount.java
// Store a simple bank account with deposit and withdrawal behavior
public class BankAccount {
    private final String owner;
    private double balance;

    /**
	 * @brief Build a bank account for one owner
	 *
	 * @param owner Account owner name
	 *
	 * @param balance Starting balance
	 */
    public BankAccount(String owner, double balance) {
        this.owner = owner;
        this.balance = balance;
    }

    /**
	 * @brief Add money to the account balance
	 *
	 * @param amount Amount to deposit
	 */
    public void deposit(double amount) {
        balance += amount;
    }

    /**
	 * @brief Attempt to withdraw money from the account
	 *
	 * @param amount Amount to withdraw
	 *
	 * @return True when the withdrawal succeeds
	 */
    public boolean withdraw(double amount) {
        // Reject withdrawals that exceed the current balance
        if (amount > balance) {
            return false;
        }

        balance -= amount;
        return true;
    }

    /**
	 * @brief Build a readable account summary
	 *
	 * @return Summary string for the account
	 */
    public String summary() {
        return owner + " has $" + balance;
    }
}
